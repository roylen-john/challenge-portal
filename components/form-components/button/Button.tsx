import React, { ReactElement } from 'react'
import { classNames } from '../../../utils/utils'

export interface iButtonProps {
  children?: React.ReactNode
  icon?: React.ReactNode
  type?: 'button' | 'submit'
  variant?: 'primary' | 'neutral'
  fullWidth?: boolean
  disabled?: boolean
  onClick?: () => void
}

const Button = ({
  children,
  icon,
  type = 'button',
  variant = 'neutral',
  fullWidth,
  disabled,
  onClick,
}: iButtonProps): ReactElement => {
  return (
    <button
      className={classNames(
        variant === 'primary'
          ? 'bg-primary hover:bg-primaryBold text-white'
          : 'bg-neutral hover:bg-neutralBold text-gray-700',
        'relative font-bold py-2 px-4 rounded focus:ring-2 focus:ring-inset focus:ring-primaryBold disabled:bg-gray-200',
        fullWidth && 'w-full'
      )}
      type={type}
      disabled={disabled}
      role="button"
      onClick={onClick}
    >
      {icon}
      {children}
    </button>
  )
}

export default Button
