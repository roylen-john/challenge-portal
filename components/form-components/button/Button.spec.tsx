import React from 'react'
import { cleanup, fireEvent, render } from '@testing-library/react'

import Button, { iButtonProps } from './Button'

const DEFAULT_PROPS: iButtonProps = {
  icon: <span>Test Icon</span>,
  type: 'button',
  variant: 'primary',
  disabled: false,
  onClick: () => {
    /** Do nothing */
  },
}

describe('Button', () => {
  afterEach(cleanup)

  test('renders correctly', () => {
    const { asFragment } = render(<Button {...DEFAULT_PROPS} />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('sets button attributes (type) correctly', () => {
    const { getByRole } = render(<Button {...DEFAULT_PROPS} />)
    const button = getByRole('button')
    expect(button).toHaveAttribute('type', DEFAULT_PROPS.type)
  })

  test('renders icon', () => {
    const { getByText } = render(<Button {...DEFAULT_PROPS} />)
    const icon = getByText('Test Icon')
    expect(icon).toBeVisible()
  })

  test('disables button', () => {
    const { getByRole } = render(<Button {...DEFAULT_PROPS} disabled />)
    const button = getByRole('button')
    expect(button).toBeDisabled()
  })

  test('calls on click when clicked', () => {
    const onClickMock = jest.fn()
    const { getByRole } = render(
      <Button {...DEFAULT_PROPS} onClick={onClickMock} />
    )
    const button = getByRole('button')
    fireEvent.click(button)
    expect(onClickMock).toHaveBeenCalled()
  })
})
