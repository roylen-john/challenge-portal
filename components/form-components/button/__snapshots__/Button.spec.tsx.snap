// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<DocumentFragment>
  <button
    class="bg-primary hover:bg-primaryBold text-white relative font-bold py-2 px-4 rounded focus:ring-2 focus:ring-inset focus:ring-primaryBold disabled:bg-gray-200"
    role="button"
    type="button"
  >
    <span>
      Test Icon
    </span>
  </button>
</DocumentFragment>
`;
